### PR TITLE
introduce `compute_indirect_effect_for`

### PR DESCRIPTION
**Pull request by cedricbatailler**
- [x] update NEWS.md
- [x] review the math
- [x] write docs
- [x] write unit tests
- [x] check code coverage

### CODE REVIEW

**Comment by cedricbatailler** [comment]
Right now, the function is not documented, but it's running.

The math behind the function should be reviewed and explained within the documentation. Briefly, the function is the mere wrapper which computes an indirect index with:

a = b_51 + b_53 * `Mod`
b = b64 + b65 * `Mod`

(see Muller et al., 2005). 

Right now, I am still unsure on how this function should relates to the `stage` arguments in `add_index.moderated_mediation`.

**Comment by cedricbatailler** [comment]
We now use an approach different that the one described [here](https://github.com/cedricbatailler/JSmediation/pull/17#issuecomment-869403962). 

The previous approach was problematic because inspecting the same place with data frames using different Moderator coding yield to different results. Even if I was not able to pinpoint what was the issue, my guess is that it has to due with the standard error associated with b_51 and b_64 which was used for the Monte Carlo simulation (see Muller et al., 2005). The new approach should fix the problem. 

Rather than computing the indirect effect by changing value of `Mod` in the equation described [here](https://github.com/cedricbatailler/JSmediation/pull/17#issuecomment-869403962), the code run a new moderated mediation with a data set where the moderator's coding was changed. 

In short, the code takes the data set used for the moderated mediation and changes `Mod`'s coding so that the value for which we want to compute the indirect effect is now `0`. Because of how you'd interpret main effect in a linear regression, b_51 and b_64 now represents the effect of the IV on the Mediator and the effect of the Mediator on the DV controlling for the effect of the IV **when Mod equals 0**. Indirect effect is computed based on the product of b_51 and b_64.

**Comment by cedricbatailler** [comment]
The good news is that the implemented method give results consistent with the `mediation` package. Numeric difference are negligible and expected anyways given the difference of approach (`mediation`: percentile bootstrap, `JSmediation`: Monte Carlo estimation). 

Another interesting observation is that `JSmediation` is more than 300% faster than `mediation`.

``` r
# install.packages("mediation")
# remotes::install_github("cedricbatailler/JSmediation", 
#                         ref = "feature-check_indirect_effect_for")
# install.packages("tictoc")

library(tidyverse)
library(JSmediation)
library(mediation)
#> Loading required package: MASS
#> 
#> Attaching package: 'MASS'
#> The following object is masked from 'package:dplyr':
#> 
#>     select
#> Loading required package: Matrix
#> 
#> Attaching package: 'Matrix'
#> The following objects are masked from 'package:tidyr':
#> 
#>     expand, pack, unpack
#> Loading required package: mvtnorm
#> Loading required package: sandwich
#> mediation: Causal Mediation Analysis
#> Version: 4.5.0
library(tictoc)

dataset <- 
  ho_et_al %>% 
  mutate(condition_c = build_contrast(condition,
                                      "Low discrimination",
                                      "High discrimination")) %>% 
  mutate(across(c(linkedfate, sdo), 
                ~ as.numeric(scale(.))))

# mediation approach
tic()
model_1 <- lm(linkedfate ~ condition_c * sdo, as.data.frame(dataset))
model_2 <- lm(hypodescent ~ condition_c * sdo + linkedfate, as.data.frame(dataset))

mediation_model <-
  mediation::mediate(model_1, model_2,    
                     covariates = list(sdo = 0), 
                     boot = TRUE,   
                     boot.ci.type = "perc", 
                     sims = 5000, 
                     treat    = "condition_c", 
                     mediator = "linkedfate")
#> Running nonparametric bootstrap
toc()
#> 62.619 sec elapsed

# JSmediation approach
tic()
JSmediation_model <-
  dataset %>% 
  mdt_moderated(DV = hypodescent,
                IV = condition_c,
                M = linkedfate,
                Mod = sdo) %>% 
  compute_indirect_effect_for(Mod = 0)
toc()
#> 0.191 sec elapsed

# results
summary(mediation_model)  
#> 
#> Causal Mediation Analysis 
#> 
#> Nonparametric Bootstrap Confidence Intervals with the Percentile Method
#> 
#> (Inference Conditional on the Covariate Values Specified in `covariates')
#> 
#>                Estimate 95% CI Lower 95% CI Upper p-value    
#> ACME             0.0920       0.0433         0.15  <2e-16 ***
#> ADE              0.0388      -0.1188         0.20   0.630    
#> Total Effect     0.1308      -0.0175         0.28   0.085 .  
#> Prop. Mediated   0.7032      -3.3441         5.79   0.085 .  
#> ---
#> Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1
#> 
#> Sample Size Used: 824 
#> 
#> 
#> Simulations: 5000
JSmediation_model
#> - type: Conditional simple mediation index (Mod = 0) 
#> - point estimate: 0.0916 
#> - confidence interval:
#>   - method: Monte Carlo (5000 iterations)
#>   - level: 0.05 
#>   - CI: [0.0425; 0.146]
```

<sup>Created on 2021-07-01 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0)</sup>

**Comment by cedricbatailler** [comment]
Current implementation also give results consistent with the `processR` package, as tested with the `processR` [shiny app](https://cardiomoon.shinyapps.io/processR/). 

image

**Comment by cedricbatailler** [comment]
Right now, the function is flagged as "experimental". This usually indicates that the API might change in the future. I'm wondering whether we should keep it that way. The alternative is a function flagged as "stable" (for now, this is implicit).

File: R/compute_indirect_effect_for.R [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+#' @param mediation_model A moderated mediation model fitted with `mdt_moderated``. 
+#' @param Mod The moderator value for which to compute the indirect effect. Must
+#'   be a numeric value, defaults to `0`.
+#' @param times Number of simulations to use to compute the Monte Carlo indirect
+#'   effect confidence interval. Must be numeric, defaults to `5000`.
+#' @param level Alpha threshold to use for the indirect effect's confidence
+#'   interval.
+#'   
+#' @details 
+#'   The approach usde by `compute_indirect_effect_for` is similar to the
```
Comment (YoannJulliard) [inline R/compute_indirect_effect_for.R]:
```suggestion
#'   The approach used by `compute_indirect_effect_for` is similar to the
```

**Review by YoannJulliard — APPROVED** [review]
